hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/DefaultS3ClientFactory.java

```diff
@@ -175,6 +163,39 @@ public S3AsyncClient createS3AsyncClient(
     return s3AsyncClientBuilder.build();
   }
 
+  private S3AsyncClient createS3CrtAsyncClient(URI uri, S3ClientCreationParameters parameters)
+      throws IOException {
+    Configuration conf = getConf();
+    String bucket = uri.getHost();
+
+    S3CrtAsyncClientBuilder s3CrtAsyncClientBuilder = S3CrtAsyncClient.builder();
+
+    AWSRegionEndpointInformation regionEndpointInformation =
+        AWSRegionEndpointResolver.getEndpointRegionResolution(parameters, conf);
+
+    if (regionEndpointInformation.getRegion() != null) {
+      s3CrtAsyncClientBuilder.region(regionEndpointInformation.getRegion());
+    }
+
+    if (regionEndpointInformation.getEndpoint() != null) {
+      s3CrtAsyncClientBuilder.endpointOverride(regionEndpointInformation.getEndpoint());
```

Comment:
Is setting both endpoint and region expected ? Won't the client throw exception in that case ?

Reply:
yeah, I haven't made any changes to the logic, just refactored it a bit. We've always had the ability for uses to set both the endpoint and the region.. I think maybe useful if you want to us a specific endpoint for a region (eg: fips enabled?) .. this whole logic does need a bit of a rethink and re-write, see: https://issues.apache.org/jira/browse/HADOOP-19470